Sources/HttpPipeline/NIO.swift

```diff
@@ -0,0 +1,135 @@
+    let bootstrap = ServerBootstrap(group: group)
+      .serverChannelOption(ChannelOptions.backlog, value: 256)
+      .serverChannelOption(reuseAddrOpt, value: 1)
+      .childChannelInitializer { channel in
+        channel.pipeline.configureHTTPServerPipeline().then {
+          channel.pipeline.add(handler: Handler(middleware))
+        }
+      }
+      .childChannelOption(ChannelOptions.socket(IPPROTO_TCP, TCP_NODELAY), value: 1)
+      .childChannelOption(reuseAddrOpt, value: 1)
+      .childChannelOption(ChannelOptions.maxMessagesPerRead, value: 1)
+
+    let serverChannel = try bootstrap.bind(host: "localhost", port: port).wait()
+    try serverChannel.closeFuture.wait()
+    try group.syncShutdownGracefully()
+  } catch {
+    fatalError(error.localizedDescription)
+  }
+}
+
+private final class Handler: ChannelInboundHandler {
+  typealias InboundIn = HTTPServerRequestPart
+
+  var request: URLRequest?
+  let middleware: Middleware<StatusLineOpen, ResponseEnded, Prelude.Unit, Data>
+
+  init(_ middleware: @escaping Middleware<StatusLineOpen, ResponseEnded, Prelude.Unit, Data>) {
+    self.middleware = middleware
+  }
+
+  func channelRead(ctx: ChannelHandlerContext, data: NIOAny) {
+    let reqPart = self.unwrapInboundIn(data)
+
+    switch reqPart {
+    case let .head(header):
+      self.request = URL(string: header.uri).map { url -> URLRequest in
+        var req = URLRequest(url: url)
+          |> \.httpMethod .~ method(from: header.method)
+          |> \.allHTTPHeaderFields .~ header.headers.reduce(into: [:]) { $0[$1.name] = $1.value }
+        req.url = req.value(forHTTPHeaderField: "Host")
+          .flatMap { URL(string: "http://" + $0 + url.absoluteString) }
+        return req
+      }
+    case let .body(bodyPart):
+      self.request = self.request.flatMap { req -> URLRequest? in
+        req.value(forHTTPHeaderField: "Content-Length")
+          .flatMap(Int.init)
+          .flatMap { bodyPart.getBytes(at: 0, length: $0) }
+          .map { req |> \.httpBody .~ Data($0) }
+      }
+    case .end:
+      guard let req = self.request else {
+        _ = ctx.channel.close()
+        return
+      }
+
+      let conn = self.middleware(connection(from: req)).perform()
+      let res = conn.response
+
+      let head = HTTPResponseHead(
+        version: .init(major: 1, minor: 1),
+        status: .init(statusCode: res.status.rawValue),
+        headers: .init(res.headers.map { ($0.name, $0.value) })
+      )
+      _ = ctx.channel.write(HTTPServerResponsePart.head(head))
+
+      var buffer = ctx.channel.allocator.buffer(capacity: res.body.count)
+      buffer.write(bytes: res.body)
+      _ = ctx.channel.write(HTTPServerResponsePart.body(.byteBuffer(buffer)))
+
+      _ = ctx.channel.writeAndFlush(HTTPServerResponsePart.end(nil)).then {
+        ctx.channel.close()
+      }
+    }
+  }
+}
+
+private func method(from method: HTTPMethod) -> String {
```

Comment:
Maybe I'm missing something, but there didn't seem to be a way to get the string representation of this thing.